README.md
# simple-audio-codec-unity

Provides simple decoding and encoding of audio codecs by [NAudio](https://github.com/naudio/NAudio) for Unity.

## Features

- Low load to the main thread of Unity
  - Runs decoding of data on a thread pool using [UniTask](https://github.com/Cysharp/UniTask) and set data to AudioClip on the main thread.
- Low memory allocation
  - Decodes WAVE file for each block of frames with array buffers.

## How to import by UnityPackageManager

```json
{
  "dependencies": {
    "com.mochineko.simple-audio-codec-unity": "https://github.com/mochi-neko/simple-audio-codec-unity.git?path=/Assets/Mochineko/SimpleAudioCodec#0.2.0",
    "com.naudio.core": "https://github.com/mochi-neko/simple-audio-codec-unity.git?path=/Assets/NAudio/NAudio.Core#0.2.0",
    "com.cysharp.unitask": "https://github.com/Cysharp/UniTask.git?path=src/UniTask/Assets/Plugins/UniTask",
    ...
  }
}
```

## Avairable codecs

- WAV
  - [x] Decoding
  - [x] Encoding
- MP3
  - [ ] Decoding
  - [ ] Encoding

## Hot to use

See [Samples](https://github.com/mochi-neko/simple-audio-codec-unity/tree/main/Assets/Mochineko/SimpleAudioCodec.Samples) or [Tests](https://github.com/mochi-neko/simple-audio-codec-unity/tree/main/Assets/Mochineko/SimpleAudioCodec.Tests). 

## Changelog

See [CHANGELOG](https://github.com/mochi-neko/simple-audio-codec-unity/blob/main/CHANGELOG.md).

## 3rd Party Notices

See [NOTICE](https://github.com/mochi-neko/simple-audio-codec-unityy/blob/main/NOTICE.md).

## License

[MIT License](https://github.com/mochi-neko/simple-audio-codec-unity/blob/main/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added 

- Add MP3 decoding with NAudio / NLayer
- Add MP3 encoding with NAudio?

## [0.2.0] - 2023-03-13

### Added

- Add WAV encoding with NAudio.

### Changed

- Refactor WAV decoding.

### Fixed

- Fix samples count of AudioClip created by WAV decoding.

## [0.1.2] - 2023-03-12

### Fixed

- Fix multi-channels decoding error of WAV.

## [0.1.1] - 2023-03-11

### Fixed

- Fix failures to decode WAV.

## [0.1.0] - 2023-03-11

### Added

- Add WAV decoding with NAudio.


Assets/Mochineko/SimpleAudioCodec/WaveDecoder.cs
#nullable enable
using System;
using System.IO;
using System.Threading;
using Cysharp.Threading.Tasks;
using NAudio.Wave;
using UnityEngine;

namespace Mochineko.SimpleAudioCodec
{
    /// <summary>
    /// A wave file decoder.
    /// </summary>
    public static class WaveDecoder
    {
        private const int RecommendedFramesCountInBlock = 1024 * 32;

        /// <summary>
        /// Decodes wave file to AudioClip with buffering a block of samplef frames.
        /// If it decodes large file at once, it uses large memories.
        /// If it decodes large file frame by frame, thread switching is overhead of process.
        /// Therefore it decodes blocks of frames.
        /// </summary>
        /// <param name="stream">Wave data stream</param>
        /// <param name="fileName">File name</param>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <param name="framesCountInBlock">Total number of frames in a frame block</param>
        /// <returns>Created AudioClip from Wave file</returns>
        /// <exception cref="Exception"></exception>
        /// <exception cref="OperationCanceledException"></exception>
        /// <exception cref="FormatException"></exception>
        /// <exception cref="InvalidDataException"></exception>
        /// <exception cref="IOException"></exception>
        /// <exception cref="ObjectDisposedException"></exception>
        public static async UniTask<AudioClip> DecodeByBlockAsync(
            Stream stream,
            string fileName,
            CancellationToken cancellationToken,
            int framesCountInBlock = RecommendedFramesCountInBlock)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (!stream.CanRead)
            {
                throw new IOException($"Cannot read stream.");
            }

            await using var reader = new WaveFileReader(stream);
            var format = reader.WaveFormat;
            if (format == null)
            {
                throw new Exception($"Wave format is null.");
            }

            // Can create AudioClip only on the main thread
            await UniTask.SwitchToMainThread(cancellationToken);

            var audioClip = AudioClip.Create(
                name: fileName,
                lengthSamples: (int)reader.SampleCount,
                channels: format.Channels,
                frequency: format.SampleRate,
                stream: false
            );
            
            await UniTask.SwitchToThreadPool();

            // total samples = channels * frames
            var samplesBuffer = new float[framesCountInBlock * format.Channels];
            // block align = bytes per frame = channels * bytes per sample
            var bytesBuffer = new byte[framesCountInBlock * format.BlockAlign];
            var sampleOffset = 0;

            while (true) 
            {
                cancellationToken.ThrowIfCancellationRequested();
                
                var readSamplesCount = reader
                    .ReadNextBlockOfFramesWithBuffer(framesCountInBlock, samplesBuffer, bytesBuffer);

                // End of file
                if (readSamplesCount == 0)
                {
                    break;
                }

                if (readSamplesCount == framesCountInBlock)
                {
                    // Can write to AudioClip only on the main thread
                    await UniTask.SwitchToMainThread(cancellationToken);
                    
                    audioClip.SetData(samplesBuffer, sampleOffset);
                }
                else
                {
                    // Set just enough length of samples
                    var justLengthSamples = new float[readSamplesCount];
                    Array.Copy(samplesBuffer, justLengthSamples, readSamplesCount);
                    
                    await UniTask.SwitchToMainThread(cancellationToken);
                    
                    audioClip.SetData(justLengthSamples, sampleOffset);
                    
                    break;
                }

                await UniTask.SwitchToThreadPool();

                sampleOffset += readSamplesCount;
            }

            // Return to the main thread
            await UniTask.SwitchToMainThread(cancellationToken);

            return audioClip;
        }
    }
}

Assets/Mochineko/SimpleAudioCodec.Tests/WaveDecoderTest.cs
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using NAudio.Wave;
using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;
using Debug = UnityEngine.Debug;

namespace Mochineko.SimpleAudioCodec.Tests
{
    [TestFixture]
    internal sealed class WaveDecoderTest
    {
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 512)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 256)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 128)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 64)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 32)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 16)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 8)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 4)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 2)]
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 1)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 512)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 256)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 128)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 64)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 32)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 16)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 8)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 4)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 2)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 1)]
        [RequiresPlayMode(true)]
        public async Task DecodeTest(string fileName, int framesCountInBlock)
        {
            var filePath = Path.Combine(
                Application.dataPath,
                "Mochineko/SimpleAudioCodec.Tests",
                fileName);

            await using var stream = File.OpenRead(filePath);

            var stopWatch = new Stopwatch();
            stopWatch.Start();

            var audioClip = await WaveDecoder.DecodeByBlockAsync(
                stream, fileName, CancellationToken.None, framesCountInBlock);

            stopWatch.Stop();
            
            audioClip.Should().NotBeNull();

            stream.Seek(0, SeekOrigin.Begin);

            await using var reader = new WaveFileReader(stream);
            var format = reader.WaveFormat;
            
            Debug.Log($"Decoding time:{stopWatch.ElapsedMilliseconds}ms with frames count in block:{framesCountInBlock} for {fileName}," +
                      $" Samples:{reader.SampleCount}, Channels:{format.Channels}");

            audioClip.samples.Should().Be((int)reader.SampleCount);
            audioClip.frequency.Should().Be(format.SampleRate);
            audioClip.channels.Should().Be(format.Channels);
            
            UnityEngine.Object.Destroy(audioClip);
        }
    }
}

Assets/Mochineko/SimpleAudioCodec.Tests/WaveEncoderTest.cs
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using NAudio.Wave;
using NUnit.Framework;
using UnityEditor;
using UnityEngine;
using UnityEngine.TestTools;
using Debug = UnityEngine.Debug;

namespace Mochineko.SimpleAudioCodec.Tests
{
    [TestFixture]
    internal sealed class WaveEncoderTest
    {
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 32)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 32)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024)]
        [RequiresPlayMode(true)]
        public async Task EncodeTest(string fileName, int framesCountInBlock)
        {
            var assetPath = Path.Combine(
                "Assets/Mochineko/SimpleAudioCodec.Tests",
                fileName);

            var audioClip = AssetDatabase.LoadAssetAtPath<AudioClip>(assetPath);
            
            await using var encoded = new MemoryStream();

            var stopWatch = new Stopwatch();
            stopWatch.Start();
            
            await WaveEncoder.EncodeByBlockAsync(
                outStream: encoded,
                audioClip: audioClip,
                cancellationToken: CancellationToken.None,
                bitsPerSample: 16,
                framesCountInBlock: framesCountInBlock);
            
            stopWatch.Stop();

            Debug.Log($"Encoding time:{stopWatch.ElapsedMilliseconds}ms with frames count in block:{framesCountInBlock} for {fileName}");

            encoded.Length.Should().NotBe(0);

            encoded.Seek(0, SeekOrigin.Begin);

            await using var reader = new WaveFileReader(encoded);
            var format = reader.WaveFormat;

            reader.SampleCount.Should().Be(audioClip.samples);
            format.SampleRate.Should().Be(audioClip.frequency);
            format.Channels.Should().Be(audioClip.channels);
        }
    }
}

Assets/Mochineko/SimpleAudioCodec/WaveFileReaderExtension.cs
#nullable enable
using System;
using System.IO;
using NAudio.Wave;
using UnityEngine;

namespace Mochineko.SimpleAudioCodec
{
    internal static class WaveFileReaderExtension
    {
        /// <summary>
        /// Reads next block of frames with buffers.
        /// Extends <see cref="NAudio.Wave.WaveFileReader.ReadNextSampleFrame()"/> to block of frames with buffer.
        /// </summary>
        /// <param name="reader"></param>
        /// <param name="framesCountInBlock"></param>
        /// <param name="samplesBuffer"></param>
        /// <param name="bytesBuffer"></param>
        /// <returns>Read samples count. Returns 0 at the end of file.</returns>
        /// <exception cref="InvalidOperationException">Validation errors</exception>
        /// <exception cref="IOException">I/O error</exception>
        /// <exception cref="NotSupportedException">Cannot read stream</exception>
        /// <exception cref="ObjectDisposedException">Stream is closed</exception>
        public static int ReadNextBlockOfFramesWithBuffer(
            this WaveFileReader reader,
            int framesCountInBlock,
            float[] samplesBuffer,
            byte[] bytesBuffer)
        {
            var format = reader.WaveFormat;
            // Encoding validation
            switch (format.Encoding)
            {
                case WaveFormatEncoding.Pcm:
                case WaveFormatEncoding.IeeeFloat:
                case WaveFormatEncoding.Extensible:
                    // n.b. not necessarily PCM, should probably write more code to handle this case
                    break;
                default:
                    throw new InvalidOperationException("Only 16, 24 or 32 bit PCM or IEEE float audio data supported");
            }

            var samplesCountInBlock = framesCountInBlock * format.Channels;
            if (samplesBuffer.Length != samplesCountInBlock)
            {
                throw new InvalidOperationException(
                    $"Samples count in block must be {samplesCountInBlock} but {samplesBuffer.Length}");
            }
            Array.Clear(samplesBuffer, 0, samplesBuffer.Length);

            var bytesCountInBlock = framesCountInBlock * format.BlockAlign;
            if (bytesBuffer.Length != bytesCountInBlock)
            {
                throw new InvalidOperationException(
                    $"Buffer bytes count must be {bytesCountInBlock} but {bytesBuffer.Length}");
            }
            Array.Clear(bytesBuffer, 0, bytesBuffer.Length);

            // Actually read bytes to buffer
            var readBytesCount = reader.Read(bytesBuffer, 0, bytesCountInBlock);

            // End of file
            if (readBytesCount == 0)
            {
                return 0;
            }

            var framesCountToRead = framesCountInBlock;
            if (readBytesCount < bytesCountInBlock)
            {
                // Recalculate frames count
                framesCountToRead = readBytesCount / format.BlockAlign;
            }

            // Decode block of frames
            var sampleOffset = 0;
            var bufferOffset = 0;
            for (var frameIndex = 0; frameIndex < framesCountToRead; frameIndex++)
            {
                for (var channel = 0; channel < format.Channels; channel++)
                {
                    samplesBuffer[sampleOffset] = DecodeSample(format, bytesBuffer, ref bufferOffset);
                    sampleOffset++;
                }
            }

            return framesCountToRead * format.Channels;
        }

        /// <summary>
        /// Decodes one sample of one channel from buffer with offset.
        /// Excluded from <see cref="NAudio.Wave.WaveFileReader.ReadNextSampleFrame()"/>.
        /// </summary>
        /// <param name="waveFormat"></param>
        /// <param name="buffer"></param>
        /// <param name="offset"></param>
        /// <returns>Decoded one sample of one channel as float</returns>
        /// <exception cref="InvalidOperationException">Invalid bits per samples of format</exception>
        private static float DecodeSample(WaveFormat waveFormat, byte[] buffer, ref int offset)
        {
            float result;
            if (waveFormat.BitsPerSample == 16)
            {
                result = BitConverter.ToInt16(buffer, offset) / 32768f;
                offset += 2;
            }
            else if (waveFormat.BitsPerSample == 24)
            {
                result = (((sbyte)buffer[offset + 2] << 16) | (buffer[offset + 1] << 8) | buffer[offset])
                         / 8388608f;
                offset += 3;
            }
            else if (waveFormat is { BitsPerSample: 32, Encoding: WaveFormatEncoding.IeeeFloat })
            {
                result = BitConverter.ToSingle(buffer, offset);
                offset += 4;
            }
            else if (waveFormat.BitsPerSample == 32)
            {
                result = BitConverter.ToInt32(buffer, offset) / (int.MaxValue + 1f);
                offset += 4;
            }
            else
            {
                throw new InvalidOperationException("Unsupported bit depth");
            }

            return result;
        }
    }
}

Assets/Mochineko/SimpleAudioCodec.Tests/WaveConsistencyTest.cs
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using NAudio.Wave;
using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;

namespace Mochineko.SimpleAudioCodec.Tests
{
    [TestFixture]
    internal sealed class WaveConsistencyTest
    {
        [TestCase("synthesis_2023_03_11_05_42_51.wav", 1024 * 32)]
        [TestCase("Alesis-Fusion-Pizzicato-Strings-C4.wav", 1024 * 32)]
        [RequiresPlayMode(true)]
        public async Task DecodeAndEncodeTest(string fileName, int framesCountInBlock)
        {
            var filePath = Path.Combine(
                Application.dataPath,
                "Mochineko/SimpleAudioCodec.Tests",
                fileName);

            await using var fileStream = File.OpenRead(filePath);

            await using var originalReader = new WaveFileReader(fileStream);
            var originalFormat = originalReader.WaveFormat;
            
            fileStream.Seek(0, SeekOrigin.Begin);

            var audioClip = await WaveDecoder.DecodeByBlockAsync(
                fileStream, fileName, CancellationToken.None, framesCountInBlock);

            await using var encoded = new MemoryStream();

            await WaveEncoder.EncodeByBlockAsync(
                outStream: encoded,
                audioClip: audioClip,
                cancellationToken: CancellationToken.None,
                bitsPerSample: 16,
                framesCountInBlock: framesCountInBlock);

            encoded.Seek(0, SeekOrigin.Begin);

            await using var encodedReader = new WaveFileReader(encoded);
            var encodedFormat = encodedReader.WaveFormat;

            encodedReader.Length.Should().Be(originalReader.Length);
            encodedReader.SampleCount.Should().Be(originalReader.SampleCount);
            encodedFormat.BitsPerSample.Should().Be(originalFormat.BitsPerSample);
            encodedFormat.Channels.Should().Be(originalFormat.Channels);
            encodedFormat.SampleRate.Should().Be(originalFormat.SampleRate);

            UnityEngine.Object.Destroy(audioClip);
        }
    }
}

Assets/Mochineko/SimpleAudioCodec.Samples/WaveDecodingSample.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Cysharp.Threading.Tasks;
using UnityEngine;
using UnityEngine.Assertions;
using Object = UnityEngine.Object;

namespace Mochineko.SimpleAudioCodec.Samples
{
    internal sealed class WaveDecodingSample : MonoBehaviour
    {
        [SerializeField] private string fileName;
        [SerializeField] private AudioSource audioSource = null;

        private AudioClip audioClip;
        
        private void Awake()
        {
            Assert.IsNotNull(audioSource);
        }

        [ContextMenu("Decode")]
        public async Task Decode()
        {
            if (audioClip != null)
            {
                Object.Destroy(audioClip);
                audioClip = null;
            }

            var cancellationToken = this.GetCancellationTokenOnDestroy();
            
            var filePath = Path.Combine(
                Application.dataPath,
                "Mochineko/SimpleAudioCodec.Tests",
                fileName);

            Stream stream;
            try
            {
                stream = File.OpenRead(filePath);
            }
            catch (Exception e)
            {
                Debug.LogException(e);
                return;
            }

            try
            {
                audioClip = await WaveDecoder.DecodeByBlockAsync(stream, fileName, cancellationToken);
                Debug.Log($"Succeeded to decode wave:{audioClip.samples}");
            }
            catch (Exception e)
            {
                Debug.LogException(e);
                return;
            }
            finally
            {
                await stream.DisposeAsync();
            }

            await UniTask.SwitchToMainThread(cancellationToken);

            audioSource.clip = audioClip;
            audioSource.PlayOneShot(audioClip);
        }
    }
}

Assets/Mochineko/SimpleAudioCodec/WaveEncoder.cs
#nullable enable
using System;
using System.IO;
using System.Threading;
using Cysharp.Threading.Tasks;
using NAudio.Utils;
using NAudio.Wave;
using UnityEngine;

namespace Mochineko.SimpleAudioCodec
{
    /// <summary>
    /// A wave file encoder.
    /// </summary>
    public static class WaveEncoder
    {
        private const int RecommendedBitsPerSample = 16;
        private const int RecommendedFramesCountInBlock = 1024 * 32;

        /// <summary>
        /// Encodes samples in AudioClip to wave file with buffering a block of sample frames.
        /// </summary>
        /// <param name="outStream">Bytes written stream</param>
        /// <param name="audioClip">Source AudioClip</param>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <param name="bitsPerSample">Number of bit per sample. It must 16(ordinary), 24 or 32</param>
        /// <param name="framesCountInBlock">Number of frames in a block</param>
        /// <exception cref="IOException"></exception>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ObjectDisposedException"></exception>
        /// <exception cref="OperationCanceledException"></exception>
        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static async UniTask EncodeByBlockAsync(
            Stream outStream,
            AudioClip audioClip,
            CancellationToken cancellationToken,
            int bitsPerSample = RecommendedBitsPerSample,
            int framesCountInBlock = RecommendedFramesCountInBlock)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (!outStream.CanWrite)
            {
                throw new IOException($"Cannot write to outStream.");
            }

            if (bitsPerSample is not (16 or 24 or 32))
            {
                throw new InvalidOperationException("Unsupported bit depth");
            }
            
            // Prevent to dispose outStream when WaveFileWriter has been disposed.
            var stream = new IgnoreDisposeStream(outStream);

            await UniTask.SwitchToMainThread(cancellationToken);

            if (audioClip.loadType != AudioClipLoadType.DecompressOnLoad)
            {
                throw new InvalidOperationException($"Only can get data from AudioClip set load type:{AudioClipLoadType.DecompressOnLoad}.");
            }

            var format = new WaveFormat(
                rate: audioClip.frequency,
                bits: bitsPerSample,
                channels: audioClip.channels);

            var totalSamplesCount = audioClip.samples * audioClip.channels;
            
            await UniTask.SwitchToThreadPool();

            await using var writer = new WaveFileWriter(stream, format);

            var samplesCountInBlock = framesCountInBlock * format.Channels;
            var blockBuffer = new float[samplesCountInBlock];
            var samplesOffset = 0;
            var framesOffset = 0;

            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();
                
                var residue = totalSamplesCount - (samplesOffset + samplesCountInBlock);
                
                Debug.Log($"Offset:{samplesOffset}");
                
                // Read full samples of a block
                if (residue > 0)
                {
                    await UniTask.SwitchToMainThread(cancellationToken);
                    
                    // A sample of AudioClip = sample frame of WAV, then specify frames count to offsetSamples
                    audioClip.GetData(blockBuffer, framesOffset);
                    
                    await UniTask.SwitchToThreadPool();
                    
                    writer.WriteSamples(blockBuffer, 0, samplesCountInBlock);

                    framesOffset += framesCountInBlock;
                    samplesOffset += samplesCountInBlock;
                }
                // Read last block that is full samples or less samples
                else
                {
                    var residueSamplesCount = totalSamplesCount - samplesOffset;
                    var residueBuffer = new float[residueSamplesCount];
                    
                    await UniTask.SwitchToMainThread(cancellationToken);
                    
                    audioClip.GetData(residueBuffer, framesOffset);
                    
                    await UniTask.SwitchToThreadPool();
                    
                    writer.WriteSamples(residueBuffer, 0, residueSamplesCount);
                    
                    await writer.FlushAsync(cancellationToken);
                    
                    // Return to the main thread
                    await UniTask.SwitchToMainThread(cancellationToken);
                    return;
                }
            }
        }
        
        /// <summary>
        /// Encodes samples to wave file into stream.
        /// </summary>
        /// <param name="outStream">Bytes written stream</param>
        /// <param name="samples">Encoded samples</param>
        /// <param name="channels">Channel count</param>
        /// <param name="sampleRate">Sampling rate</param>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <param name="bitsPerSample">Number of bit per sample. It must 16(ordinary), 24 or 32</param>
        /// <exception cref="IOException"></exception>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ObjectDisposedException"></exception>
        /// <exception cref="OperationCanceledException"></exception>
        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static async UniTask EncodeAsync(
            Stream outStream,
            float[] samples,
            int channels,
            int sampleRate,
            CancellationToken cancellationToken,
            int bitsPerSample = RecommendedBitsPerSample)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (!outStream.CanWrite)
            {
                throw new IOException($"Cannot write to outStream.");
            }

            if (bitsPerSample is not (16 or 24 or 32))
            {
                throw new InvalidOperationException("Unsupported bit depth");
            }
            
            await UniTask.SwitchToThreadPool();
            
            // Prevent to dispose outStream when WaveFileWriter has been disposed.
            var stream = new IgnoreDisposeStream(outStream);

            var format = new WaveFormat(
                rate: sampleRate,
                bits: bitsPerSample,
                channels: channels);

            await using var writer = new WaveFileWriter(stream, format);

            writer.WriteSamples(samples, 0, samples.Length);

            await writer.FlushAsync(cancellationToken);
            
            await UniTask.SwitchToMainThread(cancellationToken);
        }
    }
}